modules/pubsub/operations.go

```diff
@@ -0,0 +1,167 @@
+	}
+	
+	subject = utils.SingleLeadingTrailing(subject, "/")
+
+	// Check if the user is authorised to make this request
+	err := m.auth.IsPublishAuthorised(project, token, subject, map[string]interface{}{})
+	if err != nil {
+		return http.StatusForbidden, errors.New("You are not authorized to make this request")
+	}
+
+	// m.RLock()
+	// defer m.RUnlock()
+
+	err = m.connection.Publish(subject, &model.PubsubMsg{subject, data})
+	if err != nil {
+		return 500, err
+	} else {
+		return 200, nil
+	}
+}
+
+func (m *Module) Subscribe(project, token, clientID, subject string, cb model.PubsubCallback) (int, error) {
+	return m.QueueSubscribe(project, token, clientID, subject, clientID, cb)
+}
+
+func (m *Module) QueueSubscribe(project, token, clientID, subject, queue string, cb model.PubsubCallback) (int, error) {
+	// Exit if pubsub is not enabled
+	if !m.IsEnabled() {
+		return http.StatusNotFound, errors.New("This feature isn't enabled")
+	}
+	
+	subject = utils.SingleLeadingTrailing(subject, "/")
+
+	// Check if the user is authorised to make this request
+	err := m.auth.IsSubscribeAuthorised(project, token, subject, queue, map[string]interface{}{})
+	if err != nil {
+		return http.StatusForbidden, errors.New("You are not authorized to make this request")
+	}
+
+	m.RLock()
+	defer m.RUnlock()
+	
+	receiveChannel := make(chan *model.PubsubMsg)
+	subs, err := m.connection.QueueSubscribe(subject, queue, receiveChannel)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	subscription := pubsubSubscription{subs, receiveChannel}
+	err = m.storeSubs(subject, clientID, &subscription)
+	if err != nil {
+		subscription.unsubscribeFunc()
+		close(subscription.ch)
+		return http.StatusInternalServerError, err
+	}
+	go func() {
+		for msg := range receiveChannel {
+        	cb(msg)
+		}
+	}()
+	return http.StatusOK, nil
+}
+
+func (m *Module) Unsubscribe(clientID, subject string) (int, error) {
+	// Exit if pubsub is not enabled
+	if !m.IsEnabled() {
+		return http.StatusNotFound, errors.New("This feature isn't enabled")
+	}
+	
+	subject = utils.SingleLeadingTrailing(subject, "/")
+
+	m.Lock()
+	defer m.Unlock()
+
+	subjM, ok := m.storage.Load(subject)
+	if ok {
+		subjectMap := subjM.(sync.Map)
+		subs, ok := subjectMap.Load(clientID)
+		if ok {
+			subscription := subs.(*pubsubSubscription)
+			subjectMap.Delete(clientID)
+			if mapLen(&subjectMap) == 0 {
+				m.storage.Delete(subject)
+			}
+			(*subscription).unsubscribeFunc()
+			close((*subscription).ch)
+		}
+	}
+	return http.StatusOK, nil
+}
+
+func (m *Module) UnsubscribeAll(clientID string) (int, error) {
+	// Exit if pubsub is not enabled
+	if !m.IsEnabled() {
+		return http.StatusNotFound, errors.New("This feature isn't enabled")
+	}
+
+	// m.RLock()
+	// defer m.RUnlock()
+	
+	var err error = nil
+	status := 200
+
+	m.storage.Range(func(subject, v interface{}) bool {
+		s, e := m.Unsubscribe(clientID, subject.(string))
+		if e != nil {
```

Comment:
Ignore the error here. Try and unsubscribe for each subject

Reply:
Ok